### PR TITLE
GenAI 10.12

### PR DESCRIPTION
**Pull request by rik-bos**
Include:
* High level introduction into GenAI
* Basic capabilities / concepts & how-tos
* Conversational UI as new module
* GenAI as new module
* Updated Bedrock & OpenAI connector with dependency on GenAI
* Moved documents to create a section
* Bedrock referral page in the GenAI section

To do:

Draft:

Ready for review:

Reviewed:
* content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md // New document (Mark)
* content/en/docs/appstore/modules/genai/concepts/_index.md // (Mark)
* content/en/docs/appstore/modules/aws/amazon-bedrock.md -- added operations, included dependency on GenAI commons (Kasia)
* content/en/docs/appstore/modules/genai/concepts/rag-example-implementation.md -- minor changes; mainly moved. (Kasia)
* content/en/docs/appstore/modules/genai/_index.md (Kasia)
* content/en/docs/appstore/modules/genai/concepts/function-calling.md // Small changes to include Bedrock. (Kasia)
* content/en/docs/appstore/modules/genai/genai-commons.md // Rik to validate with the team (Kasia)
* content/en/docs/appstore/modules/genai/bedrock.md // Redirect page (Kasia)
* content/en/docs/appstore/modules/openai/_index.md //  (Dana)
* content/en/docs/appstore/modules/genai/conversational-ui.md // (Kasia)

### CODE REVIEW

**Comment by rik-bos** [comment]
content/en/docs/appstore/modules/genai/_index.md is ready for review.
Others will be added when we complete them.

File: content/en/docs/appstore/modules/aws/amazon-bedrock.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ The [Amazon Bedrock](https://marketplace.mendix.com/link/component/215042) conne
```
Comment (katarzyna-koltun-mx) [inline content/en/docs/appstore/modules/aws/amazon-bedrock.md:16]: Reviewed.

File: content/en/docs/appstore/modules/genai/concepts/rag-example-implementation.md [hunk 2]
```diff
@@ -2,7 +2,7 @@
 title: "RAG Example Implementation in the OpenAI Showcase Application"
```
Comment (katarzyna-koltun-mx) [inline content/en/docs/appstore/modules/genai/concepts/rag-example-implementation.md:2]: Reviewed.

File: content/en/docs/appstore/modules/genai/_index.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
+---
```
Comment (katarzyna-koltun-mx) [inline content/en/docs/appstore/modules/genai/_index.md:1]: Reviewed.

File: content/en/docs/appstore/modules/genai/_index.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+2. Connect any model through our GenAI connectors, or by integrating your own connector into our GenAI commons interface.
+3. Connect your own data to ground GenAI systems with data from inside your application and the rest of your IT landscape.
+
+## 2 Architecture & Components
+
+### 2.1 The Mendix Components
+
+TODO: insert diagram
+
+| Asset | Description | Studio Pro Version |
```
Comment (rik-bos) [inline content/en/docs/appstore/modules/genai/_index.md:34]: @katarzyna-koltun-mx - everywhere where this table mentions 9.24 it's actually 9.24.2 (that was still TBD), can you update this?

**Comment by MarkvanMents** [comment]
@katarzyna-koltun-mx - can you check that the URLs are consistent and have a think about what the URL structure should be?
I prefer all the docs to have the URL structure /appstore/modules/genai/<docname>, but do you think we need more distinction between specific technical solutions?

File: content/en/docs/appstore/modules/genai/concepts/_index.md [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+- **Conversational Agents:** Power chatbots and virtual assistants to interact naturally with users.
+- **Data Generation:** Generate Mendix objects by combining JSON generation & [Import mappings](/refguide/import-mapping-action/).
+
+Models such as [Anthropic Claude](/appstore/modules/aws/amazon-bedrock/#chat-completions-with-history) and [GPT-4o](/appstore/modules/openai-connector/#chatcompletions-vision) can also use (one or multiple) images as input. This will allow you to ask questions on images for use cases such as Object Recognition, Image to Text (OCR) and validating whether an image is as intended.
+
+### 1.2 What is an LLM Not?
+
+While LLMs are powerful, they are not without limitations:
+
+- **Not Conscious:** LLMs do not possess self-awareness or understanding. They generate text based on patterns in the data they were trained on.
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/genai/concepts/_index.md]: Should we say what we mean by not having "understanding"?

```suggestion
- **Not Conscious:** LLMs do not possess self-awareness or semantic knowledge (understanding). They generate text based on patterns in the data they were trained on.
```

**Comment by MarkvanMents** [comment]
Case question:
Are we going for "generative AI" as opposed to "Generative AI"?

File: content/en/docs/appstore/modules/genai/concepts/_index.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+This pattern is supported both by [OpenAI](https://platform.openai.com/docs/guides/function-calling) and [Bedrock in the Anthropic Claude v3 models](https://docs.anthropic.com/en/docs/tool-use).
+
+## 5 Agents {#agents}
+
+### 5.1 Connect to a Bedrock Agent
+
+Agents for Amazon Bedrock provides the ability to integrate autonomous agents into your application. A [Bedrock Agent](https://aws.amazon.com/bedrock/agents/) can orchestrate interactions between LLM's, various data sources and user conversations. In addtion, agents can be connected to a knowledge base to perform RAG and autonomously take actions by calling APIs that it has been equipped with.
+
+Connecting to an agent from a Mendix application can easily be done via the Amazon Bedrock Connector. For detailed instructions please refer to the [Connector documentation](/appstore/modules/aws/amazon-bedrock).
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/genai/concepts/_index.md]: Should we get rid of this section - it doesn't seem to go with the rest of the document and is incomplete.
```suggestion
```

File: content/en/docs/appstore/modules/genai/concepts/_index.md [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+<!-- TODO Update AI BOT STARTER APP link -->
+
+### 3.2 PgVector Knowledge Base with OpenAI or Bedrock {pgvectorknowledgebase}
+
+You can also use the [PgVector Knowledge Base module](/appstore/modules/pgvector-knowledge-base/) in combination with an OpenAI or Bedrock embeddings model, to maintain and use a knowledge base. Index and store your own knowledge in a more dynamic way, since you can use it for storing chunks of texts, or contents of Mendix objects.
+
+You can read further on [how to implement this here](/appstore/modules/openai-connector/rag-example-implementation/). The [OpenAI Showcase App](https://marketplace.mendix.com/link/component/220475) also contains examples on:
+* How to ask questions on a large piece of text.
+* How to use a multi-lingual database of historical tickets & resolutions to propose resolutions to the users based on historical input.
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/genai/concepts/_index.md]: This seems to be solution-specific information, and doesn't contain all the information (for example it implies you can't use RAG with OpenAI.
I think it should be removed and moved into the solution-specific documentation?

File: content/en/docs/appstore/modules/genai/concepts/_index.md [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+[Read more on this pattern](/appstore/modules/openai-connector/function-calling/) or download the [OpenAI Showcase App](https://marketplace.mendix.com/link/component/220475) for an example of allowing an LLM to answer a question like _What is the status of ticket 42?_ by providing it a microflow like _GetInformationForTicketID_.
+
+This pattern is supported both by [OpenAI](https://platform.openai.com/docs/guides/function-calling) and [Bedrock in the Anthropic Claude v3 models](https://docs.anthropic.com/en/docs/tool-use).
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/genai/concepts/_index.md]: This is not generic information and I think we should move it to the module-specific documentation.

**Review by MarkvanMents — CHANGES_REQUESTED** [review]
Reviewed concepts/_index.md
Have indicated some things I think should be removed or changed.
@rik-bos - can you have a look and see what you think?

File: content/en/docs/appstore/modules/genai/_index.md (hunk 4 above)
Comment (katarzyna-koltun-mx) [reply content/en/docs/appstore/modules/genai/_index.md:34]: Done

File: content/en/docs/appstore/modules/genai/_index.md [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+
+## 2 Architecture and Components
+
+To help you get started, the following sections list the available GenAI components and models.
+
+### 2.1 Mendix Components
+
+| Asset | Description | Studio Pro Version |
+|----------------------|------------------------------------|----------------|
+| AI Bot Starter App | Lets you kick-start the development of enterprise-grade AI chatbot experiences. For example, you can use it to create your own private enterprise-ready ChatGPT-like app. | 10.12 |
```
Comment (katarzyna-koltun-mx) [inline content/en/docs/appstore/modules/genai/_index.md]: @rik-bos Let me know when the links in line 36 and 37 are available. I also deleted the line that said "diagram here" - if you have the diagram, please send it to me and I'll add it back.

File: content/en/docs/appstore/modules/genai/concepts/function-calling.md [hunk 10]
```diff
@@ -0,0 +1,91 @@
+---
```
Comment (katarzyna-koltun-mx) [inline content/en/docs/appstore/modules/genai/concepts/function-calling.md:1]: Reviewed.

File: content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md [hunk 11]
```diff
@@ -0,0 +1,197 @@
...
+
+### 2.2 Context and Additional Information
+
+After telling the model what to do, you can include additional context. This can include:
+
+* information about the end-user of your application – for example their language, role, department, specific database records
+* context information (for example, all data related to an object the end-user is looking at)
+* knowledge coming from [Retrieval Augmented Generation (RAG)](/appstore/modules/genai/concepts/#rag)
+
+Tip: you can use export to JSON or XML to provide this information in a structured way.
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md]: Is this not just repeating section 2.4, below, or do you mean something different here?

File: content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+### 2.3 Input Data
+
+The actual input, usually from an end-user.
+
+### 2.4 Output Style
+
+You can instruct the model to format the output in a specific way. For example:
+
+* tell the model to specify its reasoning steps, or just give the answer
+* give examples of the output style you want, for example a JSON structure, if you want to import the response to  generate data or get information about the reasoning
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md]: Not sure how JSON helps you get information about the reasoning.

**Review by MarkvanMents — COMMENTED** [review]
Hi @rik-bos 
I have now reviewed the two documents which I was assigned. Gen AI Concepts and Prompt Engineering. Please have a look at the changes, and also the comments:

I left some suggested changes in my [review of the GenAI Concepts document](https://github.com/mendix/docs/pull/7920#pullrequestreview-2132770694).

I also have a couple of comments on the Prompt Engineering document.

File: content/en/docs/appstore/modules/genai/concepts/_index.md (hunk 8 above)
Comment (rik-bos) [reply content/en/docs/appstore/modules/genai/concepts/_index.md]: @MarkvanMents - In general I want to say that both OpenAI and Bedrock supports this pattern. Does that make sense? We can phrase it in a different way.

For example:

This pattern is supported both by [OpenAI](https://platform.openai.com/docs/guides/function-calling) and [Bedrock](https://docs.anthropic.com/en/docs/tool-use).

or

See [Function Calling](/appstore/modules/openai-connector/function-calling/) for more information on ReAct with OpenAI and Bedrock. You can see...

File: content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md (hunk 12 above)
Comment (rik-bos) [reply content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md]: Suggestion: 
> * give examples of the output style you want, for example a JSON structure, if you want to import the response to generate data or get structured information on the different steps performed in the propmt

Background:
When a prompt instructs a more complex process containing of multiple steps, you can instruct it to store substeps in JSON so you can import both the answer as well as the intermediate substeps.

Or in the case of this prompt you can ask the model to explain what is factual and what not:

```
When should I use Office 365 vs Google Docs. Give me a list of arguments and whether they are a preference or based on facts. Also give an example. Respond in a JSON array.
```

Response:
```
...
{
"argument": "Advanced features and functionalities",
"preference_or_fact": "fact",
"example": "Office 365 typically offers more advanced features such as better formatting options, more powerful data analysis tools in Excel, and advanced presentation features in PowerPoint."
},
{
"argument": "Ease of use",
"preference_or_fact": "preference",
"example": "Some users find Google Docs to be more intuitive and user-friendly, while others prefer the familiar interface of Office 365."
},
...
```

File: content/en/docs/appstore/modules/genai/concepts/_index.md (hunk 5 above)
Comment (rik-bos) [reply content/en/docs/appstore/modules/genai/concepts/_index.md]: Yeah, that looks good.

File: content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md (hunk 11 above)
Comment (rik-bos) [reply content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md]: Yes, this is different. If a user want's to add information to the context (so the prompt / LLM input) they can use Export to JSON/XML to present that information to the LLM in a structured way.

2.4 is that the LLM responds in JSON and that the reponse can be parsed with JSON afterwards.
For that you need to provide an example of the JSON in the prompt to make sure that the JSON structure is as expected.

File: content/en/docs/appstore/modules/genai/openai/_index.md [hunk 13]
```diff
@@ -0,0 +1,817 @@
...
+3. Open the downloaded file and follow the installation steps. Remember the installation path. Usually, this should be something like `C:/Program Files/Eclipse Adoptium/jdk-11.0.22.7-hotspot`.
+4. After the installation has finished, restart your computer if prompted.
+5. Open Studio Pro and go to **Edit** > **Preferences** > **Deployment** > **JDK directory**. Click **Browse** and select the folder with the new JDK version you just installed. This should be the folder containing the *bin* folder. Save your settings by clicking **OK**.
+6. Run the project and execute the action that threw the above-mentioned exception earlier.
+    1. You might get an error saying `FAILURE: Build failed with an exception. The supplied javaHome seems to be invalid. I cannot find the java executable.`. In this case, verify that you have selected the correct JDK directory containing the updated JDK version. You may also need to update Gradle. To do this, go to **Edit** > **Preferences** > **Deployment** > **Gradle directory**. Click **Browse** and select a newer Gradle version from the Mendix folder. In this case, we replaced `grade-7.6` with `gradle-7.6.3`. Save your settings by clicking **OK**.
```
Comment (dbreseman) [inline content/en/docs/appstore/modules/genai/openai/_index.md]: > In this case, we replaced `grade-7.6` with `gradle-7.6.3`.

Could you clarify this? Could this just read "Make sure to use `gradle-7.6.3`."? Or is it that the user needs whatever the newest available version is in the Mendix folder?

File: content/en/docs/appstore/modules/genai/concepts/_index.md (hunk 6 above)
Comment (MarkvanMents) [reply content/en/docs/appstore/modules/genai/concepts/_index.md]: Updated after discussion
25745caee2806e6ccb8e3ee6799ce84cdc769080

File: content/en/docs/appstore/modules/genai/concepts/_index.md (hunk 7 above)
Comment (MarkvanMents) [reply content/en/docs/appstore/modules/genai/concepts/_index.md]: Updated after discussion.
25745caee2806e6ccb8e3ee6799ce84cdc769080

File: content/en/docs/appstore/modules/genai/concepts/_index.md (hunk 8 above)
Comment (MarkvanMents) [reply content/en/docs/appstore/modules/genai/concepts/_index.md]: Updated after discussion.
25745caee2806e6ccb8e3ee6799ce84cdc769080

File: content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md (hunk 12 above)
Comment (MarkvanMents) [reply content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md]:
```suggestion
* give examples of the output style you want, for example a JSON structure, if you want to use a structured response to generate data or get structured information about intermediate steps taken, or decisions made, in coming up with a final response to a prompt.
```

**Review by MarkvanMents — COMMENTED** [review]
z

File: content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+### 2.2 Context and Additional Information
+
+After telling the model what to do, you can include additional context. This can include:
+
+* information about the end-user of your application – for example their language, role, department, specific database records
+* context information – for example, all data related to an object the end-user is looking at
+* knowledge coming from [Retrieval Augmented Generation (RAG)](/appstore/modules/genai/concepts/#rag)
+
+Tip: you can export to JSON or XML to provide this information in a structured way.
```
Comment (MarkvanMents) [inline content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md]:
```suggestion
Tip: you can provide information in a JSON or XML structure to ensure the information is presented in a consistent way. From Mendix apps, you can use [Export Mappings](/refguide/export-mappings/) to create JSON structures and [Export XML Documents](/howto/integration/export-xml-documents/) to create XML structures.
```

File: content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md (hunk 11 above)
Comment (MarkvanMents) [reply content/en/docs/appstore/modules/genai/concepts/prompt-engineering.md]: See https://github.com/mendix/docs/pull/7920/files#r1651322302 for a suggested update.

**Review by MarkvanMents — CHANGES_REQUESTED** [review]
@rik-bos 
I have just made two final suggestions to resolve the discussion about JSON and XML structures. Hope they are good. Let me know and they can be committed and I can approve the PR.

File: content/en/docs/appstore/modules/genai/genai-commons.md [hunk 15]
```diff
@@ -0,0 +1,446 @@
+---
```
Comment (katarzyna-koltun-mx) [inline content/en/docs/appstore/modules/genai/genai-commons.md:1]: Reviewed. Post-release will want to add a more detailed configuration section.

File: content/en/docs/appstore/modules/genai/bedrock.md [hunk 16]
```diff
@@ -0,0 +1,19 @@
+---
```
Comment (katarzyna-koltun-mx) [inline content/en/docs/appstore/modules/genai/bedrock.md:1]: Reviewed.

File: content/en/docs/appstore/modules/genai/openai/_index.md (hunk 13 above)
Comment (dbreseman) [reply content/en/docs/appstore/modules/genai/openai/_index.md]: @rik-bos What do you think about this?

Comment (rik-bos) [reply content/en/docs/appstore/modules/genai/openai/_index.md]: > https://docs.mendix.com/refguide/install/#offline suggests to use:
> For Mendix 10.10.0 and above: gradle-8.5-bin.zip
> For Mendix 10 versions below 10.10.0: gradle-7.6.3-bin.zip

I'd indeed change this to something like:

You might get an error saying `FAILURE: Build failed with an exception. The supplied javaHome seems to be invalid. I cannot find the java executable.`. In this case, verify that you have selected the correct JDK directory containing the updated JDK version. You may also need to update Gradle. To do this, go to **Edit** > **Preferences** > **Deployment** > **Gradle directory**. Click **Browse** and select a newer Gradle version from the Mendix folder.
For Mendix 10.10.0 and above use gradle 8.5.
For Mendix 10 versions below 10.10.0 use gradle 7.6.3.

_NB: Can you check with the runtime tech writer whether there should be some generic piece somewhere on this? I don't belong it ultimately should be documented here. Perhaps only referred to from here._

Comment (dbreseman) [reply content/en/docs/appstore/modules/genai/openai/_index.md]: Thanks. I'll implement this change and follow up with @MarkvanMents about longer-term options for documenting this.

**Review by MarkvanMents — APPROVED** [review]
`concepts/_index` and `prompt-engineering` are good to go from my end.

File: content/en/docs/appstore/modules/genai/conversational-ui.md [hunk 17]
```diff
@@ -0,0 +1,186 @@
+---
```
Comment (katarzyna-koltun-mx) [inline content/en/docs/appstore/modules/genai/conversational-ui.md:1]: Reviewed but the Configuration section is very sparse and should be updated post-release.

Comment (rik-bos) [reply content/en/docs/appstore/modules/genai/conversational-ui.md:1]: Agreed. @liamsommer-mx - can you include this in the docs enrichment story that we should include some screenshots and more text on how to do this?

**Pull request merged**